File: Backend/PingMeChat.CMS.Entities/Feature/Call.cs
using PingMeChat.CMS.Entities.Interfaces;
using PingMeChat.CMS.Entities.Users;
using System;
using System.Collections.Generic;
using System.ComponentModel.DataAnnotations;
using System.Linq;
using System.Text;
using System.Threading.Tasks;

namespace PingMeChat.CMS.Entities.Feature
{
    // Call entity
    public class Call : AuditableBaseEntity
    {
        public string ChatId { get; set; }
        public CallStatus CallStatus { get; set; }
        public CallType CallType { get; set; }
        public DateTime StartTime { get; set; } 
        public DateTime? EndTime { get; set; }
        public virtual Chat Chat { get; set; }
        public virtual ICollection<CallParticipant> CallParticipants { get; set; }
    }

    public enum CallStatus
    {
        InProgress,
        Completed,
        Missed
    }

    public enum CallType
    {
        Audio,
        Video
    }


}

File: Frontend/pingmechat_ui/lib/domain/models/call.dart
import 'package:pingmechat_ui/domain/models/callparticipant.dart';
import 'package:pingmechat_ui/domain/models/chat.dart';

enum CallStatus {
  inProgress,
  completed,
  missed,
}

enum CallType {
  audio,
  video,
}

class Call {
  String chatId;
  CallStatus callStatus;
  CallType callType;
  DateTime startTime;
  DateTime? endTime;
  Chat chat;
  List<CallParticipant> callParticipants;

  Call({
    required this.chatId,
    required this.callStatus,
    required this.callType,
    required this.startTime,
    this.endTime,
    required this.chat,
    required this.callParticipants,
  });

  factory Call.fromJson(Map<String, dynamic> json) {
    return Call(
      chatId: json['chatId'],
      callStatus: CallStatus.values[json['callStatus']],
      callType: CallType.values[json['callType']],
      startTime: DateTime.parse(json['startTime']),
      endTime: json['endTime'] != null ? DateTime.parse(json['endTime']) : null,
      chat: Chat.fromJson(json['chat']),
      callParticipants: (json['callParticipants'] as List)
          .map((participant) => CallParticipant.fromJson(participant))
          .toList(),
    );
  }

  Map<String, dynamic> toJson() {
    return {
      'chatId': chatId,
      'callStatus': callStatus.index,
      'callType': callType.index,
      'startTime': startTime.toIso8601String(),
      'endTime': endTime?.toIso8601String(),
      'chat': chat.toJson(),
      'callParticipants': callParticipants.map((participant) => participant.toJson()).toList(),
    };
  }
}
